File: src/core/detector.cpp
#include "asc/core/detector.hpp"

#include <algorithm>
#include <stdexcept>

namespace asc {

DebouncedDetector::DebouncedDetector(const DetectorSettings settings) : settings_(settings) {
    if (settings_.threshold < 0.0 || settings_.threshold > 1.0 ||
        settings_.matches_required == 0 || settings_.mismatches_required == 0) {
        throw std::invalid_argument("invalid detector settings");
    }
}

void DebouncedDetector::configure(const DetectorSettings settings) {
    if (settings.threshold < 0.0 || settings.threshold > 1.0 || settings.matches_required == 0 || settings.mismatches_required == 0)
        throw std::invalid_argument("invalid detector settings");
    settings_ = settings;
    snapshot_.consecutive_matches = 0;
    snapshot_.consecutive_mismatches = 0;
}

DetectionSnapshot DebouncedDetector::update(const double similarity, const bool capture_valid, const TimePoint now) {
    snapshot_.similarity = std::clamp(similarity, 0.0, 1.0);
    snapshot_.measured_at = now;
    if (!capture_valid) {
        snapshot_.state = DetectionState::reference_missing;
        snapshot_.consecutive_matches = 0;
        snapshot_.consecutive_mismatches = 0;
        return snapshot_;
    }
    if (snapshot_.similarity >= settings_.threshold) {
        snapshot_.consecutive_mismatches = 0;
        ++snapshot_.consecutive_matches;
        if (snapshot_.consecutive_matches >= settings_.matches_required) {
            snapshot_.state = DetectionState::matching;
        }
    } else {
        snapshot_.consecutive_matches = 0;
        ++snapshot_.consecutive_mismatches;
        if (snapshot_.consecutive_mismatches >= settings_.mismatches_required) {
            snapshot_.state = DetectionState::not_matching;
        }
    }
    return snapshot_;
}

void DebouncedDetector::reset(const DetectionState state) {
    snapshot_ = {};
    snapshot_.state = state;
}

} // namespace asc

File: crates/core/src/detector.rs
use crate::DetectionState;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct DetectorSettings {
    pub threshold: f64,
    pub matches_required: u32,
    pub mismatches_required: u32,
}

impl Default for DetectorSettings {
    fn default() -> Self {
        Self {
            threshold: 0.98,
            matches_required: 5,
            mismatches_required: 3,
        }
    }
}

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct DebouncedDetector {
    settings: DetectorSettings,
    state: DetectionState,
    matches: u32,
    mismatches: u32,
    similarity: f64,
}

impl DebouncedDetector {
    pub fn new(settings: DetectorSettings) -> Self {
        Self {
            settings,
            ..Self::default()
        }
    }
    pub fn state(&self) -> DetectionState {
        self.state
    }
    pub fn similarity(&self) -> f64 {
        self.similarity
    }
    pub fn counters(&self) -> (u32, u32) {
        (self.matches, self.mismatches)
    }
    pub fn reset(&mut self) {
        self.state = DetectionState::Unknown;
        self.matches = 0;
        self.mismatches = 0;
    }

    pub fn update(&mut self, similarity: f64, capture_valid: bool) -> DetectionState {
        self.similarity = similarity;
        if !capture_valid || !similarity.is_finite() {
            self.matches = 0;
            self.mismatches = 0;
            self.state = DetectionState::ReferenceMissing;
        } else if similarity >= self.settings.threshold {
            self.matches = self.matches.saturating_add(1);
            self.mismatches = 0;
            if self.matches >= self.settings.matches_required {
                self.state = DetectionState::Matching;
            }
        } else {
            self.mismatches = self.mismatches.saturating_add(1);
            self.matches = 0;
            if self.mismatches >= self.settings.mismatches_required {
                self.state = DetectionState::NotMatching;
            }
        }
        self.state
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn applies_five_three_debounce() {
        let mut detector = DebouncedDetector::new(DetectorSettings::default());
        for _ in 0..4 {
            assert_eq!(detector.update(0.99, true), DetectionState::Unknown);
        }
        assert_eq!(detector.update(0.99, true), DetectionState::Matching);
        for _ in 0..2 {
            assert_eq!(detector.update(0.1, true), DetectionState::Matching);
        }
        assert_eq!(detector.update(0.1, true), DetectionState::NotMatching);
    }
}
